### DISCUSSION TITLE
Struggling to see how supabase makes dev easier/better...

### DISCUSSION THREAD

**Discussion by flxmrk** [supabase/supabase#11529] [General]
So I've spent the last few months trying to build something out of Supabase... and I'm well aware anything new presents a learning curve... but I feel I've reached the point to conclude its not helpful at all.

Let's be real, very few apps will be happy living inside RLS.  Then supabase requires you to spend all your time trying to write complex plpgsql policies that help you achieve the restricted access your app requires, and things end up so convoluted you end up falling back on security definer RPCs to do something as simple as reading some fields off data the user doesn't own.

Then I think, oh, EDGE FUNCTIONS, that must be the way to do some server-side coding, breaking out of the RPC barriers, but still keeping my secrets secure... nope... i think they changed that in some recent big shift toward "no superuser"?  I found no advantage to edge functions that helps me get access to the data my users would need.  Sure, if you have an api key to twilio and need to send a message without exposing your key, edge function.  But edge function to help you with supabase data, nope, might as well just keep that in your client side app, no advantage there.

In the end, building a nice standard REST server-side with some mongo is far faster and cleaner than supabase (or even firebase etc) can be IMHO. I would end up feeling far more comfortable with the security of my data than I do now with all these convoluted policies written to do what used to be very simple things.

Sure I love the IDEA of just querying my data directly from my apps, until I actually try to USE it in a real app... then I figure out its messy, with logic in illogical places thats unsearchable from my IDE... mmmm nope.

I guess I can see its attraction to devs without sufficient backend or sysadmin experience... but if you know how to handle those things... why would you choose services like supa/firebase/etc?

**Comment by BernieSumption** [comment]
Have you tried using the service role key within an edge function in order to bypass RLS? That sounds like what you're looking for.

A potential source of confusion is that supabase _allows_ you to write server-side code that runs as the same user that is calling the function, and is subject to the same RLS policies. This is useful for situations when, as you say, you want to protect your twilio API key but don't want to accidentally allow the user to bypass RLS.

But there's nothing to stop you from creating an API client using your service role key and getting full read-write access. You'd then be responsible for implementing your own checks in code to prevent inappropriate access.

> Let's be real, very few apps will be happy living inside RLS.

Writing RLS policies to completely cover your app's data access needs probably would be cumbersome in many apps. I have simple policies like "anyone can read, users can only write their own data" for user profile data. For more complex use cases I use the service role key within a trusted server environment (I have a NodeJS backend, Edge Functions should work just as well).

> Then I think, oh, EDGE FUNCTIONS, that must be the way to do some server-side coding, breaking out of the RPC barriers, but still keeping my secrets secure... nope... i think they changed that in some recent big shift toward "no superuser"?

If you want to use a regular postgres client and SQL, then either restore superuser status to the postres user [as described here](https://github.com/supabase/supabase/discussions/2495) or create a new user with the appropriate privileges.

**Reply by flxmrk** [reply]
Thanks for your response.  I didn't know about any "Service role" that could be used in edge functions, to be honest, the documentation on this project leaves much to be desired.  When you read about edge functions its very specific scenario they document, and that is basically the same as code i would put into the app.

I think its scary to give read to public on user profiles.  Basically means any bad actor could pull your whole database of users.  But if you hope to give other users the names of the other users, now you have to jump through policy craziness to do something that normally is quite simple to provide securely.

So I understand from your message that its possible to break out of the RLS with server-side edge functions, but in the end my original thoughts remain - why bother?  When you know how to spin up laravel and build a beautiful complex model complete with fine-grained permissions using sanctum, what benefit would firebase/supabase have?

Thus my current conclusion - these kinds of things are for people with no backend capability or for tinkers.  And even if for some reason you wanted realtime graphql or data subscriptions, why not spin up hasura with postgres?

**Reply by BernieSumption** [reply]
I agree that the documentation can be better.

Supabase offers a superset of features found in other Postgres hosts. Personally I find the automatic API and authentication tied to RLS to be extremely useful and it saves me a lot of time especially in the early stages of a project. Using a different Postgres host I'd also need a different authentication provider, and would need to roll my own authorisation.

I can totally understand that if your preference is not to use these features then supabase is "just another Postgres host" and you may prefer a different one.

But supabase *is* a postgres host, which means that you can do anything on it that you can do on any other host, including spinning up hasura if that's your thing.
